drivers/serial/uart_cc13xx_cc26xx.c

```diff
@@ -313,6 +493,22 @@ static int uart_cc13xx_cc26xx_init_0(struct device *dev)
 {
 	int ret;
 
+#ifdef CONFIG_DEVICE_POWER_MANAGEMENT
+	get_dev_data(dev)->pm_state = DEVICE_PM_ACTIVE_STATE;
+#endif
+
+#ifdef CONFIG_SYS_POWER_MANAGEMENT
+	get_dev_data(dev)->rx_constrained = false;
```

Comment:
Are these the right defaults when interrupt-driven UART is not enabled?  It looks like the constraints are only added as a result of enabling interrupts.  (IMO it'd be fine to require interrupt-driven UART if power management is enabled.)

Reply:
My concern is that only that `SLEEP_2` (standby mode) is disabled if and only if the constrained variables are true (which only happens when interrupts are enabled, meaning the UART is active), and the comments suggest this is because UART won't work in standby mode.  So by inspection it seems that non-interrupt use of UART would fail if the system goes into standby.  (I'm concerned more about losing characters on reception because the UART is shut down; polled output will probably work because the system can't sleep during the spin-wait.)

At the moment the sleep states are default-enabled, though #21118 changes them to default-disabled.  (I expect that to get merged this development cycle.)

Reply:
Ahh, now I get it, thanks for the clarification - I was thinking of the poll-out. It is indeed a problem on the poll-in. The tricky part here is that even if CONFIG_UART_INTERRUPT_DRIVEN is thrown to enable interrupt mode, the non-interrupt mode functions are still available to be called and are not disabled. So users that are calling poll-in today would still suffer character lost.

Several alternatives I can think of:
1. Inside uart_cc13xx_cc26xx_poll_in, disable sleep state 2 if it is enabled, and issue a warning message to the log: "Using poll-in in uart driver automatically disables sleep state 2. Suggest to use interrupt-driven API if possible for better power savings". This would ensure code such as this would work regardless of of the value of SYS_PM_MIN_RESIDENCY_SLEEP_2:
```
        while (size) {
                u8_t c;
                res = uart_poll_in(tty->uart_dev, &c);

                /* Avoid 100% busy-polling, and yet try to process bursts
                 * of data without extra delays.
                 */
                if (res == -1) {
                        k_sleep(K_MSEC(1));
                }
        }
```

2. Change all the call sites in Zephyr to use interrupt-driven API instead of poll-in (e.g. tty.c), and make CONFIG_UART_INTERRUPT_DRIVEN disable the poll-in API. (will break existing code that relies on both at once)

3. Document in device-specific docs that poll-in may lose characters on this platform if sleep state 2 is enabled when PM is turned on (not very user-friendly for those who turn on PM without reading the docs)

Any thoughts on these? I am leaning towards #1 since it minimizes disruptions to other platforms, allows applications to at least work (albeit with sub-optimal PM), and removes the chance of someone scratching their head wondering why characters are getting lost.

Reply:
I'm not sure.  I'm still trying to figure out the complexities of the UART driver.  I'm aware that poll-based UART is already a problem for power management (`uart_poll_out` at the API level doesn't guarantee the data is flushed before the call completes, so it can be lost generally even if you've patched things to make it work in this driver).  I was also thinking of cases where an enabled peripheral while sleeping would put something into the RX buffer where it could be picked up by a later `uart_poll_in()`.

The short-term solution I was thinking is that you can't put the UART into sleep mode when `CONFIG_UART_INTERRUPT_DRIVEN=n` but UART is enabled, so you would initialize the `?x_constrained` variables here to `true` and add the constraint at the same time.

The long-term solution is to come up with an API-level understanding of how power management affects UART, both poll and interrupt-driven.

Reply:
I tend to concur that the UART driver API would benefit from some adjustment to make power management easier. For example, if I compare this with the UART driver API in TI-RTOS, I see that they follow an open/close model (so that the device and its associated power domains can be powered only when in opened state at runtime, whereas in Zephyr we need to assume the device would be powered as soon as it is initialized and the CPU is in active state). They adhere to the concept of a UART transaction, so that the standby state in only locked for the duration of the transfer. As soon as the transaction is done the CPU can go into standby. So they do not have the issue we are seeing here, where the expectation is that the UART stays receptive to input even after the poll-in function has returned with its transaction essentially completed.

Afaik, `CONFIG_UART_INTERRUPT_DRIVEN=y` does not imply that all UART transfers would be interrupt-driven. Polling is still supported, so I don't think we can make the inference to set the constraints simply based on that flag.

For simplicity, I think I'll just document this issue in the short term, and we should aim for a UART API refinement in the longer term as you suggest.